tcmur_cmd_handler.c

```diff
@@ -288,6 +292,608 @@ static int handle_write_verify(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+	 * byte header.
+	 *
+	 * The data length in CDB should be equal to tdll + sdll + inline_dl
+	 * + parameter list header length
+	 */
+	if (data_length < (XCOPY_HDR_LEN + tdll + sdll + inline_dl)) {
+		tcmu_err("Illegal parameter list length: data length from CDB is %u,"
+			 " but here the length is %u\n",
+			 data_length, tdll + sdll + inline_dl);
+		ret = tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					  ASC_PARAMETER_LIST_LENGTH_ERROR,
+					  NULL);
+		goto err;
+	}
+
+	tcmu_dbg("Processing XCOPY with tdll: %hu sdll: %u inline_dl: %u\n",
+		 tdll, sdll, inline_dl);
+
+	/*
+	 * Parse the segment descripters and for now we only support block
+	 * -> block type.
+	 *
+	 * The max seg_desc number support is 1(see RCR_OP_MAX_SG_DESC_COUNT)
+	 */
+	seg_desc = par + XCOPY_HDR_LEN + tdll;
+	ret = xcopy_parse_segment_descs(seg_desc, xcopy, sdll, sense);
+	if (ret != SAM_STAT_GOOD)
+		goto err;
+
+	/*
+	 * Parse the target descripter
+	 *
+	 * The max seg_desc number support is 2(see RCR_OP_MAX_TARGET_DESC_COUNT)
+	 */
+	tgt_desc = par + XCOPY_HDR_LEN;
+	ret = xcopy_parse_target_descs(dev, xcopy, tgt_desc, tdll, sense);
+	if (ret != SAM_STAT_GOOD)
+		goto err;
+
+	if (tcmu_get_dev_block_size(xcopy->src_dev) !=
+	    tcmu_get_dev_block_size(xcopy->dst_dev)) {
+		tcmu_err("The block size of src dev %u != dst dev %u\n",
+			 tcmu_get_dev_block_size(xcopy->src_dev),
+			 tcmu_get_dev_block_size(xcopy->dst_dev));
+		ret = tcmu_set_sense_data(sense, NOT_READY,
+					  ASC_LOGICAL_UNIT_COMMUNICATION_FAILURE,
+					  NULL);
+		goto err;
+	}
+
+	return SAM_STAT_GOOD;
+
+err:
+	free(par);
+
+	return ret;
+}
+
+static void handle_xcopy_null_cbk(struct tcmu_device *dev,
+				  struct tcmulib_cmd *cmd,
+				  int ret)
+{
+}
+
+static void handle_xcopy_cbk(struct tcmu_device *dev,
+			     struct tcmulib_cmd *cmd,
+			     int ret)
+{
+	struct xcopy *xcopy = cmd->cmdstate;
+
+	aio_command_finish(dev, cmd, ret);
+	free(xcopy);
+}
+
+static int xcopy_work_fn(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+{
+	struct tcmur_handler *rhandler = tcmu_get_runner_handler(dev);
+	uint32_t block_size = tcmu_get_dev_block_size(dev);
+	struct xcopy *xcopy = cmd->cmdstate;
+	struct tcmu_device *src_dev = xcopy->src_dev, *dst_dev = xcopy->dst_dev;
+	uint64_t src_lba = xcopy->src_lba, dst_lba = xcopy->dst_lba, end_lba;
+	unsigned short lba_cnt = xcopy->lba_cnt, copy_lbas, max_lbas, copied_lba_cnt = 0;
+	uint32_t max_sectors, src_max_sectors, dst_max_sectors;
+	struct iovec iovec;
+	size_t iov_cnt;
+	int ret;
+
+	end_lba = src_lba + lba_cnt;
+
+	src_max_sectors = tcmu_get_attribute(src_dev, "hw_max_sectors");
+	dst_max_sectors = tcmu_get_attribute(dst_dev, "hw_max_sectors");
+
+	max_sectors = min(src_max_sectors, dst_max_sectors);
+	max_sectors = min(max_sectors, (uint32_t)XCOPY_MAX_SECTORS);
+	max_lbas = min(max_sectors, ((uint32_t)(~0U)));
+
+	iovec.iov_len = min(lba_cnt, max_lbas) * block_size;
+	iovec.iov_base = calloc(1, iovec.iov_len);
+	iov_cnt = 1;
+
```

Comment:
Ok.

Not sure if it will be easier, but for testing you can also use linux and the sg utils tools, sg_write_same, sg_xcopy, sg_compare_and_write, etc.